lib/reservation_manager.rb

```diff
@@ -0,0 +1,136 @@
+      else
+        room_number = available_rooms(check_in, check_out).first
+        # reservation_rate = reservation_cost(check_in, check_out)
+
+        new_reservation = Hotel::Reservation.new(room_number,
+                                                 check_in,
+                                                 check_out)
+      end
+      @reservations << new_reservation
+
+      return @reservations
+    end
+
+    def reservations_by_date(start_date, end_date)
+      s_date = Date.parse(start_date)
+      e_date = Date.parse(end_date)
+      reservations_by_date = []
+      @reservations.each do |reservation|
+        if e_date > reservation.check_in && e_date <= reservation.check_out
+          reservations_by_date << reservation
+        elsif s_date >= reservation.check_in && s_date < reservation.check_out
+          reservations_by_date << reservation
+        elsif s_date <= reservation.check_in && e_date >= reservation.check_out
+          reservations_by_date << reservation
+
+          #Right now checkout day is not included because you can check-in on that day
+        end
+      end
+
+      return reservations_by_date
+    end
+
+    def blocked_rm_by_date(start_date, end_date)
+      s_date = Date.parse(start_date)
+      e_date = Date.parse(end_date)
+      blocks_by_date = []
+      @block.each do |room|
+        if e_date > room.check_in && e_date <= room.check_out
+          blocks_by_date << room
+        elsif s_date >= room.check_in && s_date < room.check_out
+          blocks_by_date << room
+        elsif s_date <= room.check_in && e_date >= room.check_out
+          blocks_by_date << room
+
+          #Right now checkout day is not included because you can check-in on that day
+        end
+      end
+
+      return blocks_by_date
+    end
+
+    def available_rooms(start_date, end_date)
+      unavail_rooms = reservations_by_date(start_date, end_date).map do |reservation|
+        reservation.room_number
+      end
+
+      unavail_block_rms = blocked_rm_by_date(start_date, end_date).map do |block_room|
+        block_room.room_number
+      end
+
+      all_unavail_rooms = unavail_rooms + unavail_block_rms
+
+      all_avail_rooms = all_rooms.reject { |rm_num| all_unavail_rooms.include? rm_num }
+
+      return all_avail_rooms
+    end
+
+    def make_block(total_rooms_in_block, check_in, check_out, block_id)
+      if total_rooms_in_block > 5
+        raise ArgumentError, "Blocks have a max of 5 rooms"
+      else
+        room_num_assign = available_rooms(check_in, check_out).take(total_rooms_in_block)
+
+        room_num_assign.each do |room|
+          room_number = room
+
+          block_reservation = Hotel::Reservation.new(room_number,
+                                                     check_in,
+                                                     check_out, block_id: block_id)
+
+          @block << block_reservation
+        end
+      end
+      return @block
```

Comment:
Much like the pattern I mentioned above, just wanted to say that it may make sense to follow a pattern of returning the new `block_reservation` instead